crates/cast/bin/cmd/wallet/list.rs

```diff
@@ -0,0 +1,147 @@
+use clap::Parser;
+use eyre::Result;
+
+use foundry_common::{fs, types::ToAlloy};
+use foundry_config::Config;
+use foundry_wallets::{multi_wallet::MultiWalletOptsBuilder, WalletSigner};
+
+/// CLI arguments for `cast wallet list`.
+#[derive(Clone, Debug, Parser)]
+pub struct ListArgs {
+    /// List all the accounts in the keystore directory.
+    /// Default keystore directory is used if no path provided.
+    #[clap(long, default_missing_value = "", num_args(0..=1), help_heading = "List local accounts")]
+    dir: Option<String>,
+
+    /// List accounts from a Ledger hardware wallet.
+    #[clap(long, short, help_heading = "List Ledger hardware wallet accounts")]
+    ledger: bool,
+
+    /// List accounts from a Trezor hardware wallet.
+    #[clap(long, short, help_heading = "List Trezor hardware wallet accounts")]
+    trezor: bool,
+
+    /// List accounts from AWS KMS.
+    #[clap(long, help_heading = "List AWS KMS account")]
+    aws: bool,
+
+    /// List all configured accounts.
+    #[clap(long, help_heading = "List all accounts")]
+    all: bool,
+}
+
+impl ListArgs {
+    pub async fn run(self) -> Result<()> {
+        // list local accounts as files in keystore dir, no need to unlock / provide password
+        if self.dir.is_some() || self.all || (!self.ledger && !self.trezor && !self.aws) {
+            self.list_local_senders().await?;
+        }
+
+        // Create options for multi wallet - ledger, trezor and AWS
+        let list_opts = MultiWalletOptsBuilder::default()
+            .ledger(self.ledger || self.all)
+            .mnemonic_indexes(Some(vec![0]))
+            .trezor(self.trezor || self.all)
+            .aws(self.aws || self.all)
+            .interactives(0)
+            .build()
+            .expect("build multi wallet");
+
+        // max number of senders to be shown for ledger and trezor signers
+        let max_senders = 3;
+
+        // List ledger accounts
+        match list_opts.ledgers().await {
+            Ok(signers) => {
+                self.list_senders(signers.unwrap_or_default(), max_senders, "Ledger").await?
+            }
+            Err(e) => {
+                if !self.all {
+                    println!("{}", e)
+                }
+            }
+        }
+
+        // List Trezor accounts
+        match list_opts.trezors().await {
+            Ok(signers) => {
+                self.list_senders(signers.unwrap_or_default(), max_senders, "Trezor").await?
+            }
+            Err(e) => {
+                if !self.all {
+                    println!("{}", e)
+                }
+            }
+        }
+
+        // List AWS accounts
+        match list_opts.aws_signers().await {
+            Ok(signers) => {
+                self.list_senders(signers.unwrap_or_default(), max_senders, "AWS").await?
+            }
+            Err(e) => {
+                if !self.all {
+                    println!("{}", e)
+                }
+            }
+        }
+
```

Comment:
would be nice if we could reduce duplication here via macro/fn

Reply:
yep, good point, changed in https://github.com/foundry-rs/foundry/pull/7123/commits/c4ce461125b7da4357f10c0a10e8b252039b6222